src/storage-client/src/controller/collection_mgmt.rs

```diff
@@ -288,46 +129,201 @@ impl CollectionManager {
     /// - If the collection closed.
     pub(super) async fn append_to_collection(&self, id: GlobalId, updates: Vec<(Row, Diff)>) {
         if !updates.is_empty() {
-            let (tx, _rx) = oneshot::channel();
-            self.tx.send((id, updates, tx)).await.expect("rx hung up");
+            let appender = self.monotonic_appender(id).expect("id to exist");
+            match appender.append(updates).await {
+                Ok(()) => (),
+                // There's nothing we can do if we're shutting down.
+                Err(StorageError::ShuttingDown(_)) => (),
+                Err(e) => panic!("unexpected error when appending to collection {id}, err: {e:?}"),
+            }
         }
     }
 
     /// Returns a [`MonotonicAppender`] that can be used to monotonically append updates to the
     /// collection correlated with `id`.
-    pub(super) fn monotonic_appender(&self, id: GlobalId) -> MonotonicAppender {
-        MonotonicAppender {
-            id,
-            tx: self.tx.clone(),
-        }
+    pub(super) fn monotonic_appender(
+        &self,
+        id: GlobalId,
+    ) -> Result<MonotonicAppender, StorageError> {
+        let guard = self.collections.lock().expect("CollectionManager panicked");
+        let tx = guard
+            .get(&id)
+            .map(|(tx, _)| tx.clone())
+            .ok_or(StorageError::IdentifierMissing(id))?;
+
+        Ok(MonotonicAppender { tx })
     }
 }
 
+/// Spawns a [`tokio::task`] that will continuously bump the upper for the specified collection,
+/// and append data that is sent via the provided [`mpsc::Sender`].
+///
+/// TODO(parkmycar): One day if we want to customize the tick interval for each collection, that
+/// should be done here.
+/// TODO(parkmycar): Maybe add prometheus metrics for each collection?
+fn write_task<T>(
+    id: GlobalId,
+    write_handle: persist_handles::PersistWriteWorker<T>,
+    now: NowFn,
+) -> (WriteChannel, WriteTask)
+where
+    T: Timestamp + Lattice + Codec64 + From<EpochMillis> + TimestampManipulation,
+{
+    let (tx, mut rx) = mpsc::channel(CHANNEL_CAPACITY);
+
+    let handle = mz_ore::task::spawn(
+        || format!("CollectionManager-write_task-{id}"),
+        async move {
+            let mut interval = tokio::time::interval(DEFAULT_TICK);
+
+            'run: loop {
+                tokio::select! {
+                    // Prefer sending actual updates over just bumping the upper, because sending
+                    // updates also bump the upper.
+                    biased;
+
+                    // Pull as many queued updates off the channel as possible.
+                    cmd = rx.recv_many(CHANNEL_CAPACITY) => {
+                        if let Some(batch) = cmd {
```

Comment:
if its `None` we should shut down (in addition to being able to be aborted

Reply:
Good call! Updated